roles/metrics-export.lua

```diff
@@ -1,17 +1,274 @@
+    end
+
+    if not http_handlers[endpoint.format] then
+        error("http endpoint 'format' must be one of: " ..
+              http_supported_formats_str .. ", got " .. endpoint.format, 4)
+    end
+end
+
+local function validate_http_node(node)
+    if type(node) ~= "table" then
+        error("http configuration node must be a table, got " .. type(node), 3)
+    end
+
+    local _, _, err = parse_listen(node.listen)
+    if err ~= nil then
+        error("failed to parse http 'listen' param: " .. err, 3)
+    end
+
+    node.endpoints = node.endpoints or {}
+    if type(node.endpoints) ~= "table" then
+        error("http 'endpoints' must be a table, got " .. type(node.endpoints), 3)
+    end
+    if not is_array(node.endpoints) then
+        error("http 'endpoints' must be an array, not a map", 3)
+    end
+    for _, endpoint in ipairs(node.endpoints) do
+        validate_http_endpoint(endpoint)
+    end
+
+    for i, ei in ipairs(node.endpoints) do
+        local pathi = remove_side_slashes(ei.path)
+        for j, ej in ipairs(node.endpoints) do
+            if i ~= j then
+                local pathj = remove_side_slashes(ej.path)
+                if pathi == pathj then
+                    error("http 'endpoints' must have different paths", 3)
+                end
+            end
+        end
+    end
 end
 
-local function apply()
+local httpd_servers = nil
+
+local function validate_http(conf)
+    if conf ~= nil and type(conf) ~= "table" then
+        error("http configuration must be a table, got " .. type(conf), 2)
+    end
+    conf = conf or {}
 
+    if not is_array(conf) then
+        error("http configuration must be an array, not a map", 2)
+    end
+
+    for _, http_node in ipairs(conf) do
+        validate_http_node(http_node)
+    end
+
+    for i, nodei in ipairs(conf) do
+        local hosti, porti, erri = parse_listen(nodei.listen)
+        assert(erri == nil) -- We should already successfully parse the URI.
+        for j, nodej in ipairs(conf) do
+            if i ~= j then
+                local hostj, portj, errj = parse_listen(nodej.listen)
+                assert(errj == nil) -- The same.
+                if hosti == hostj and porti == portj then
+                    error("http configuration nodes must have different listen targets", 2)
+                end
+            end
+        end
+    end
 end
 
-local function stop()
+local function apply_http(conf)
+    for _, node in ipairs(conf) do
+        if #(node.endpoints or {}) > 0 then
+            local host, port, err = parse_listen(node.listen)
+            if err ~= nil then
+                error("failed to parse URI: " .. err, 2)
+            end
+
+            local httpd = http_server.new(host, port)
+            httpd_servers = httpd_servers or {}
+            table.insert(httpd_servers, httpd)
+
+            for _, endpoint in ipairs(node.endpoints) do
+                httpd:route({
```

Comment:
Cartridge role had also cleaned up routes in case they were removed from configuration. It's a bit tricky since tarantool/http itself does not have the proper API. (Maybe the best way to implement it in a separate PR.)

https://github.com/tarantool/cartridge-metrics-role/blob/ad2b99dae13c3b9270ce113e5ae913a94a88d42d/cartridge/roles/metrics.lua#L48-L59

Reply:
Sure, added.